Automattic/wp-calypso
# Jetpack Settings: Use data-layer based implementation (alternate lyrics)

Alternative to #23393. Apologies, I'm afraid I can't reasonably split this up any more. (There are still some naming inconsistencies left for cleanup PRs since renaming would make this PR even bigger -- see bottom of PR desc.)

### This PR:

* On the API surface (query components, actions):
  * Rewires the `WrapSettingsForm` HOC to use the data-layer based `saveJetpackSettings` action from `state/jetpack-onboarding/actions` instead of the old `updateSettings` from `state/jetpack/settings/actions`.
  * Changes the `PublishingTools` settings section to use the data-layer based `regeneratePostByEmail ` action from `state/jetpack-onboarding/actions` instead of the old counterpart from `state/jetpack/settings/actions`.
  * Removes the old `QueryJetpackSettings` component, and replaces it with what was previously named `QueryJetpackOnboardingSettings`.
* At implementation level (reducers):
  * Removes the Redux state subtree that was previously called `state.jetpackOnboarding` from the root state level, and implants it instead as `state.jetpack.settings` (not changing the directory name however).
* On the API surface (selectors):
  * Re-implements the `getJetpackSettings` selector to access `state.jetpack.settings` instead of `state.jetpack.settings.items`. (Data were previously split into `items` and `requests`, but thanks to the `data-layer`'s request tracking, we no longer need the `requests` subtree.)
  * Re-implements JPO selectors such as `getJetpackOnboardingSettings` (and those based upon it) to use the `getJetpackSettings` selector.

### Testing Instructions:
* Test everything that's potentially affected by these changes:
* JPO: Try all combinations, logged out and in, connected and not.
* JP Settings: 
  * Navigate to `calypso.localhost:3000/settings/general/<YourJPsite>`.
    * Change settings and verify that changes are being applied correctly.
  * `regeneratePostByEmail`:
    * Navigate to `calypso.localhost:3000/settings/writing/<YourJPsite>`.
    * Scroll down to 'Publishing Tools'. Make sure that 'Publish posts by sending an email' is switched on.
    * Press 'Regenerate address', and verify that after a short while, the 'Email Address' field above is changed to a new, randomly generated email address.
  * Verify that changing the Akismet key still works (you might need a paid plan, not sure):
    * Navigate to `calypso.localhost:3000/settings/security/<yourJPsite>`
    * Scroll down to the 'Spam filtering' section
    * Expand the collapsed `Card` that reads 'Your site is protected from spam`
    * Copy the current (working) API key somewhere (e.g. to your text editor)
    * Change the API key to something invalid, e.g. `deadbeef`
    * Click 'Save Settings'
    * Verify that there's *one* error notice saying 'There was a problem saving your changes. Please try again.'. (Making sure there's only *one* such error confirms that https://github.com/Automattic/wp-calypso/pull/23797#pullrequestreview-109618265 is fixed)
    * Saving those changes, and there's a 'Please enter a valid Antispam API key.' warning below the input field.
    * Change the field back to its previous value, click 'Save Settings'.

### Remaining tasks for future PRs:
* Remove `client/state/jetpack/settings` directory, replace by `client/state/jetpackOnboarding` (need to keep `utils.js` tho!)
* Rename action types.
